### DISCUSSION TITLE
How to properly make schema optional on an existing schema?

### DISCUSSION THREAD

**Discussion by ReaganM02** [colinhacks/zod#3261] [Q&A] [answered]
Good day, I struggled on this part. 

I have a schema object which has a name and email props as an example, and I am using this schema on the other functions as well so I cannot manually add optional method 
```   
const schema = z.object({
  name: z.string(),
  email: z.string()
})
```

I found this, and it works
```    
function makeSchemaOptional<T extends z.ZodObject<any>> (schema: T) {
  return schema.partial()
}
```

The problem with the `makeSchemaOptional` function is it loses the typings when validating like this.
```
const optionalSchema = makeSchemaOptional(schema)
const validate = optionalSchema.safeParse({ name: 'hello' })

if (validate.success) {
  console.log(validate.data)
} else {
  console.log(validate.error.issues)
}
```
Normally, the `validate.data` will show the available props, which is the `name` and the `email` in this example. But, currently, it does not. How do I fix that? 

Stackblitz
https://stackblitz.com/edit/nuxt-starter-v28iab?file=app.vue

Thank you in advance.

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const schema = z.object( {
    name: z.string(),
    email: z.string()
} )

const optionalSchema = schema.partial()
const validate = optionalSchema.safeParse( { name: 'hello' } )

if ( validate.success ) {
    type Data = typeof validate.data
    // type Data = {
    //     name?: string | undefined
    //     email?: string | undefined
    // }
}
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by ReaganM02** [reply]
Hi, thanks for the response.   What I meant is something like this
![image](https://github.com/colinhacks/zod/assets/21333078/c434b367-3cdb-43f2-81bc-9e71c3573929)

Currently, it does not show any typings.
![image](https://github.com/colinhacks/zod/assets/21333078/735eddd5-779c-4fd6-a24f-ba1c4b54cbe5)

**Comment by JacobWeisenburger** [comment, marked as answer]
```ts
const schema = z.object( {
    name: z.string(),
    email: z.string()
} )

const optionalSchema = schema.partial()
const validate = optionalSchema.safeParse( { name: 'hello' } )

if ( validate.success ) {
    validate.data.
}
```

![image](https://github.com/colinhacks/zod/assets/31667350/68a6f099-41b3-4637-b91e-fc05ed6fd675)

**Reply by ReaganM02** [reply]
That's way simpler and understandable.  Thank you so much!!

**Reply by JacobWeisenburger** [reply]
If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
